src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -4345,32 +4545,109 @@ private uint GetTupleValEscape(ImmutableArray<BoundExpression> elements, uint sc
+        /// expressions that can be stored into the implicit receiver. That means arguments
+        /// passed to an indexer for example only matter if they can escape into the receiver
+        /// as a stored field.
+        /// </summary>
         private uint GetValEscapeOfObjectInitializer(BoundObjectInitializerExpression initExpr, uint scopeOfTheContainingExpression)
         {
             var result = CallingMethodScope;
-            foreach (var expression in initExpr.Initializers)
+            foreach (var expr in initExpr.Initializers)
+            {
+                var exprResult = GetValEscapeOfObjectMemberInitializer(expr, scopeOfTheContainingExpression);
+                result = Math.Max(result, exprResult);
+            }
+
+            return result;
+        }
+
+        private uint GetValEscapeOfObjectMemberInitializer(BoundExpression expr, uint scopeOfTheContainingExpression)
+        {
+            uint result;
+            if (expr.Kind == BoundKind.AssignmentOperator)
             {
-                if (expression.Kind == BoundKind.AssignmentOperator)
+                var assignment = (BoundAssignmentOperator)expr;
+                var rightEscape = assignment.IsRef
+                    ? GetRefEscape(assignment.Right, scopeOfTheContainingExpression)
+                    : GetValEscape(assignment.Right, scopeOfTheContainingExpression);
+
+                var left = (BoundObjectInitializerMember)assignment.Left;
+                result = left.MemberSymbol switch
                 {
-                    var assignment = (BoundAssignmentOperator)expression;
-                    var rightValEscape = assignment.IsRef
-                        ? GetRefEscape(assignment.Right, scopeOfTheContainingExpression)
-                        : GetValEscape(assignment.Right, scopeOfTheContainingExpression);
+                    PropertySymbol { IsIndexer: true } indexer => getIndexerEscape(indexer, left, rightEscape),
+                    PropertySymbol property => getPropertyEscape(property, rightEscape),
+                    _ => rightEscape
+                };
+            }
+            else
+            {
+                result = GetValEscape(expr, scopeOfTheContainingExpression);
+            }
 
-                    result = Math.Max(result, rightValEscape);
+            return result;
 
-                    var left = (BoundObjectInitializerMember)assignment.Left;
-                    result = Math.Max(result, GetValEscape(left.Arguments, scopeOfTheContainingExpression));
+            uint getIndexerEscape(
+                PropertySymbol indexer,
+                BoundObjectInitializerMember expr,
+                uint rightEscapeScope)
+            {
+                Debug.Assert(expr.AccessorKind != AccessorKind.Unknown);
+                var methodInfo = MethodInfo.Create(indexer, expr.AccessorKind);
+                Debug.Assert(methodInfo.Method is not null);
```

Comment:
>Debug.Assert(methodInfo.Method is not null); [](http://example.com/codeflow?start=16&length=44)

We're asserting that `Method is not null` here in `getIndexerEscape`, but we're handling `Method is null` in `getPropertyEscape`. Consider using the same approach in both (handle `Method is null`?) unless there is a reason the two cases are distinct.